plausible/analytics
# Edit goals with display names

### Changes

This PR introduces new capabilities of:

- editing goals
- setting up display names for goals

![image](https://github.com/user-attachments/assets/9304ff2a-86fb-4bf8-a4d5-1422c26a65df)

![image](https://github.com/user-attachments/assets/50451f5a-d69a-496c-ac3b-2a938c4a5f12)

The display name is meant to decouple event names (used in tracking integration) from what dashboard/site settings users see. Especially useful in case of long/cryptic page paths, allowing marketing-oriented users quickly identify their meaning.

It's also used in the Stats/Sites API as the primary filtering identifier, replacing conditional naming used thus far. The API docs will need an update, a relevant PR will be referenced here.

By design, it's not possible to change the goal type when editing. So for example a custom event cannot be turned into a revenue goal or a pageview. We only allow editing initially provided set of data, otherwise goal deletion is recommended.

#### Migration

By default, all the pre-existing goals are migrated, such that:

- pageviews' display names become "Visit #{page_path}"
- custom events' display names identical to event name

Migrations extracted to: #4425 and #4426 

#### UI changes

##### LiveView

The PR updates LiveView for listing and editing goals too. The form is now conditionally updated, depending on whether a `goal_id` has been passed (which effectively means edition attempt). For that reason, the optimization implemented via `PlausibleWeb.Live.Components.Modal` may not be fully applicable - the form is initially preloaded for new goal insertion, but on edit, it's updated via websocket messages to skip rendering irrelevant bits.

The optimizations mentioned were initially done to speed up modal loading in poor latency conditions. Those can be simulated locally by invoking  `liveSocket.enableLatencySim(500)` in the web inspector console, after navigating to goals site settings area.

There is one quirky bit there - applying `phash2` to DOM identifiers, so that AlpineJS picks up the changes made behind the scenes by LV. I couldn't find another way to do that. Without it, after navigating to goal settings and selecting the edit action first, the form modal won't be rendered correctly.

##### Minor UI changes

For convenience, in edit mode, clicking the display name for the first time will select the whole text input contents to allow quicker corrections.

The goals list now contains a whole lot of new information, so the list layout has been changed to accommodate them, at the same time aiming to skip display redundant info. For example, in case of pageview goal "/long_url" with display name "Visit /long_url", just the display name is shown, whilst in case of "/long_url" with display name "Pageview Alias", both will be visible.

#### Plugins API note

There should be no breaking changes other than displaying revenue goals differently in the Plugins API (used by our WordPress plugin) - the display name no longer contains currency, so "Purchase (EUR)" becomes "Purchase". This has been cleared with Daan and should cause no regressions with existing WordPress installations. 

#### How to test locally?

Re-seed the database with `mix ecto.reset`, visit goal settings, add & edit goals. Wherever goal names were used thus far, the display name should appear (dashboard, funnel settings).

#### What bugs I haven't solved yet

- [ ] ~~When editing a goal on chromium-based browsers, holding LMB to select the ComboBox dropdown contents and dragging outside of the modal window, and then releasing the LMB, triggers "click away" behaviour effectively closing the modal~~. https://github.com/alpinejs/alpine/discussions/807#discussioncomment-92461
- ~~[ ] In goal settings list, on Safari (that I don't have access to), the tooltips from `title=` attribute are not appearing on hover. Reason unknown yet. Those tooltips reportedly don't work in existing Shields sections as well.~~ not a bug, filed improvement issue separately

### Tests
- [ ] This PR does not require tests

### Changelog
- [x] Entry has been added to changelog
- [ ] This PR does not make a user-facing change

### Documentation
- [x] [Docs](https://github.com/plausible/docs) have been updated
- [ ] This change does not need a documentation update

### Dark mode
- [x] The UI has been tested both in dark and light mode
- [ ] This PR does not change the UI
